### PR TITLE
gnrc_ipv6_ext_opt: initial import

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This provides header parsing for IPv6 hop-by-hop and destination options according to [RFC 8200].

[RFC 8200]: https://tools.ietf.org/html/rfc8200#section-4.2

### Testing procedure
I provided a test application using `scapy`:

```sh
make -C tests/gnrc_ipv6_ext_opt/ flash
sudo BOARD=samr21-xpro make -C tests/gnrc_ipv6_ext_opt/ test
```

### Issues/PRs references
First step for a new adaptation for #7231

### CODE REVIEW

**Comment by miri64** [comment]
Forgot to commit the most important part... The actual module :sweat_smile:

File: sys/include/net/gnrc/ipv6/ext/opt.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#include <stdint.h>
+
+#include "net/gnrc/pkt.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Processes all options with in an IPv6 option header
```
Comment (aabadie) [inline sys/include/net/gnrc/ipv6/ext/opt.h]:
```suggestion
 * @brief   Processes all options within an IPv6 option header
```

File: tests/gnrc_ipv6_ext_opt/Makefile [hunk 2]
```diff
@@ -0,0 +1,39 @@
+DEVELHELP := 1
```
Comment (aabadie) [inline tests/gnrc_ipv6_ext_opt/Makefile]:
```suggestion
DEVELHELP = 1
```

File: tests/gnrc_ipv6_ext_opt/main.c [hunk 3]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2015-2018 Freie Universität Berlin
```
Comment (aabadie) [inline tests/gnrc_ipv6_ext_opt/main.c]:
```suggestion
 * Copyright (C) 2020 Freie Universität Berlin
```

File: tests/gnrc_ipv6_ext_opt/main.c [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       Tests extension header handling of gnrc stack.
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Takuo Yonezawa <Yonezawa-T2@mail.dnp.co.jp>
+ * @author      Martine S. Lenders <m.lenders@fu-berlin.de>
```
Comment (aabadie) [inline tests/gnrc_ipv6_ext_opt/main.c]:
```suggestion
 * @author      Martine S. Lenders <m.lenders@fu-berlin.de>
```

File: tests/gnrc_ipv6_ext_opt/tests/01-run.py [hunk 5]
```diff
@@ -0,0 +1,619 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2018 Freie Universität Berlin
```
Comment (aabadie) [inline tests/gnrc_ipv6_ext_opt/tests/01-run.py]:
```suggestion
# Copyright (C) 2020 Freie Universität Berlin
```

File: sys/include/net/ipv6/ext/opt.h [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_ipv6_ext_opt IPv6 destination and hop-by-hop options
+ * @ingroup     net_ipv6_ext
+ * @brief       Definitions for IPv6 destination and hop-by-hop extension
+ *              options extension headers
```
Comment (aabadie) [inline sys/include/net/ipv6/ext/opt.h]:
```suggestion
 *              options headers
```

File: tests/gnrc_ipv6_ext_opt/main.c [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       Tests extension header handling of gnrc stack.
```
Comment (aabadie) [inline tests/gnrc_ipv6_ext_opt/main.c]:
```suggestion
 * @brief       Tests IPv6 hop-by-hop and destination option extension header handling of gnrc stack.
```

File: tests/gnrc_ipv6_ext_opt/main.c (hunk 3 above)
Comment (miri64) [reply tests/gnrc_ipv6_ext_opt/main.c]: This is basically copied over from `tests/gnrc_ipv6_ext`, so I would like to keep copyright and authorship. I'l update the year range in the copyright note to 2015-2020 though.

File: tests/gnrc_ipv6_ext_opt/main.c (hunk 4 above)
Comment (miri64) [reply tests/gnrc_ipv6_ext_opt/main.c]: See https://github.com/RIOT-OS/RIOT/pull/13275#discussion_r388428026

File: tests/gnrc_ipv6_ext_opt/tests/01-run.py (hunk 5 above)
Comment (miri64) [reply tests/gnrc_ipv6_ext_opt/tests/01-run.py]: I'll update the year range. Most of the helper functions were copied from `tests/gnrc_ipv6_ext`.

File: sys/include/net/ipv6/ext/opt.h (hunk 6 above)
Comment (miri64) [reply sys/include/net/ipv6/ext/opt.h]: They are extension headers that provide options, so I'll remove the first "extension" to result in 

```
 * @brief       Definitions for IPv6 destination and hop-by-hop options
 *              extension headers
```

File: tests/gnrc_ipv6_ext_opt/main.c (hunk 7 above)
Comment (miri64) [reply tests/gnrc_ipv6_ext_opt/main.c]: I'll leave the "hop-by-hop and destination". It should be clear without it what is meant ;-).

**Comment by miri64** [comment]
Addressed comments

File: sys/include/net/ipv6/ext/opt.h [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+ */
+/**
+ * @brief   mask to decode action from type
+ */
+#define IPV6_EXT_OPT_ACTION_MASK            (0xc0)
+/**
+ * @brief   skip over this option and continue processing the header
+ */
+#define IPV6_EXT_OPT_ACTION_SKIP            (0x00)
+#define IPV6_EXT_OPT_ACTION_DISC            (0x40)  /**< discard the packet */
```
Comment (benpicco) [inline sys/include/net/ipv6/ext/opt.h:65]: How does this work?
AFAIU those options are already present in the header of the received header - surely there will be no option to tell the receiver should just discard the packet.
So is this indication for an option that is not supported and thus the packet should be discarded?

Comment (miri64) [reply sys/include/net/ipv6/ext/opt.h:65]: The link in the comment explains how this works: If the handling of an option is not supported by the router, it should handle the packet in accordance to the action bits. So yes, options that indicate that the packet should be discarded, if the first is not supported, the router discards the packet. Those action bits are intrinsic to the option's ID (see [IANA registry](https://www.iana.org/assignments/ipv6-parameters/ipv6-parameters.xhtml#ipv6-parameters-2)), so these bits can't just be picked by the sender at will, but are bound to the options they add to the packet.

File: sys/include/net/ipv6/ext/opt.h [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+#define IPV6_EXT_OPT_PDM            (0x0FU) /**< Performance and Diagnostic Metrics */
+/** @} */
+
+/**
+ * @name Processing actions
+ * @see [RFC 8200, section 4.2](https://tools.ietf.org/html/rfc8200#section-4.2)
+ * @{
+ */
+/**
+ * @brief   mask to decode action from type
```
Comment (benpicco) [inline sys/include/net/ipv6/ext/opt.h:58]: I'd insert a quote from the RFC here

>  The Option Type identifiers are internally encoded such that their highest-order 2 bits specify the action that must be taken if the processing IPv6 node does not recognize the Option Type.

Comment (miri64) [reply sys/include/net/ipv6/ext/opt.h:62]: Done.

**Comment by miri64** [comment]
Squashed.

**Comment by benpicco** [comment]
flake8 has a nitpick.

**Comment by miri64** [comment]
Fixed and squashed all static test failures (not just `flake8` ;-))

**Comment by benpicco** [comment]
What's left after this to implement non-storing mode?

**Comment by miri64** [comment]
The hop-by-hop option needs to be implemented still.

**Comment by miri64** [comment]
See https://github.com/miri64/RIOT/tree/gnrc_rpl_opt/feat/initial for my initial work on that.

**Pull request merged**